internal/pkg/peer/blocksprovider/bft_deliverer.go

```diff
@@ -0,0 +1,263 @@
+
+	fetchSources     []*orderers.Endpoint
+	fetchSourceIndex int
+	fetchErrorsC     chan error
+
+	blockReceiver *BlockReceiver
+
+	// TODO here we'll have a CensorshipMonitor component that detects block censorship.
+	// When it suspects censorship, it will emit an error to this channel.
+	monitorErrorsC chan error
+}
+
+func (d *BFTDeliverer) Initialize() {
+	d.requester = NewDeliveryRequester(
+		d.ChannelID,
+		d.Signer,
+		d.TLSCertHash,
+		d.Dialer,
+		d.DeliverStreamer,
+	)
+}
+
+func (d *BFTDeliverer) DeliverBlocks() {
+	var err error
+
+	d.lastBlockSourceIndex = -1
+	d.lastBlockTime = time.Now()
+	d.nextBlockNumber, err = d.Ledger.LedgerHeight()
+	if err != nil {
+		d.Logger.Error("Did not return ledger height, something is critically wrong", err)
+		return
+	}
+	d.Logger.Infof("Starting DeliverBlocks on channel `%s`, block height=%d", d.ChannelID, d.nextBlockNumber)
+
+	// select an initial source randomly
+	d.refreshSources()
+
+FetchAndMonitorLoop:
+	for {
+		// The backoff duration is doubled with every failed round.
+		// A failed round is when we had moved through all the endpoints without success.
+		// If we get a block successfully from a source, or endpoints are refreshed, the failure count is reset.
+		if count := d.getFetchFailureCounter(); count > 0 {
+			rounds := uint(count)
+			if l := len(d.fetchSources); l > 0 {
+				rounds = uint(count / len(d.fetchSources))
+			}
+
+			dur := backOffDuration(2.0, rounds, bftMinBackoffDelay, bftMaxBackoffDelay)
+			backOffSleep(dur, d.DoneC)
+		}
+
+		// assign other endpoints to the monitor
+
+		// start a block fetcher and a monitor
+		// a buffered channel so that the fetcher goroutine can send an error and exit w/o waiting for it to be consumed.
+		d.fetchErrorsC = make(chan error, 1)
+		source := d.fetchSources[d.fetchSourceIndex]
+		go d.FetchBlocks(source)
+
+		// TODO start a censorship monitor
+
+		// wait for fetch  errors, censorship suspicions, or a stop signal.
+		select {
+		case <-d.DoneC:
+			break FetchAndMonitorLoop
+		case errFetch := <-d.fetchErrorsC:
+			if errFetch != nil {
+				switch errFetch.(type) {
+				case *errStopping:
+					// nothing to do
+				case *errRefreshEndpoint:
+					// get new endpoints and reassign fetcher and monitor
+					d.refreshSources()
+					d.resetFetchFailureCounter()
+				default:
+					d.fetchSourceIndex = (d.fetchSourceIndex + 1) % len(d.fetchSources)
+					d.incFetchFailureCounter()
+				}
+				// TODO can it be nil?
+			}
+		case errMonitor := <-d.monitorErrorsC:
+			// TODO until we implement the censorship monitor this nil channel is blocked
+			if errMonitor != nil {
+				d.Logger.Warningf("Censorship suspicion: %s", err)
+				// TODO close the block receiver, increment the index
+				// TODO
+			}
+			// TODO can it be nil?
+		}
+	}
+
+	// clean up everything because we are closing
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	d.blockReceiver.Stop()
+	// TODO stop the monitor
```

Comment:
why stop the monitor from here?

Reply:
The BFTDeliverer is work in progress. It will be delt with here: https://github.com/hyperledger/fabric/issues/4348